src/main/sensors/battery.c

```diff
@@ -228,24 +230,35 @@ void batteryUpdatePresence(void)
 static void batteryUpdateVoltageState(void)
 {
     // alerts are currently used by beeper, osd and other subsystems
+    static uint32_t lastVoltageOkMillis;
     switch (voltageState) {
         case BATTERY_OK:
             if (voltageMeter.filtered <= (batteryWarningVoltage - batteryConfig()->vbathysteresis)) {
-                voltageState = BATTERY_WARNING;
+                if (cmp32(millis(), lastVoltageOkMillis) >= batteryConfig()->vbatDurationForWarning * 100) {
+                    voltageState = BATTERY_WARNING;
+                }
+            } else {
+                lastVoltageOkMillis = millis();
             }
             break;
 
         case BATTERY_WARNING:
             if (voltageMeter.filtered <= (batteryCriticalVoltage - batteryConfig()->vbathysteresis)) {
-                voltageState = BATTERY_CRITICAL;
-            } else if (voltageMeter.filtered > batteryWarningVoltage) {
-                voltageState = BATTERY_OK;
+                if (cmp32(millis(), lastVoltageOkMillis) >= batteryConfig()->vbatDurationForCrit * 100) {
+                    voltageState = BATTERY_CRITICAL;
+                }
+            } else {
+                if (voltageMeter.filtered > batteryWarningVoltage) {
+                    voltageState = BATTERY_OK;
+                }
+                lastVoltageOkMillis = millis();
             }
             break;
 
         case BATTERY_CRITICAL:
             if (voltageMeter.filtered > batteryCriticalVoltage) {
                 voltageState = BATTERY_WARNING;
+                lastVoltageOkMillis = millis();
```

Comment:
Using this here is very misleading, as the voltage is not 'ok'. Maybe call it `lastVoltageChangeMs`. (In general, variables used to keep milliseconds are called `...Ms` in Betaflight.